### PR TITLE
feat(s3-deployment): add `deployedBucket` attribute for sequencing

### PR DESCRIPTION
**Pull request by rix0rrr**
This allows referencing the bucket in a way that ensures the
deployment has happened before dependent resources are created.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by BryanPan342** [comment]
@rix0rrr would we consider this a breaking change? bc the asset id's change as a result of this! So some people who run tests might have their systems break.

**Comment by rix0rrr** [comment]
> @rix0rrr would we consider this a breaking change? bc the asset id's change as a result of this! So some people who run tests might have their systems break.

No. Snapshot testing w/ asset hashes is something people shouldn't be doing.

**Comment by rix0rrr** [comment]
I can't approve this since I opened the PR. So we need @skinny85 in the game 🙂

File: packages/@aws-cdk/aws-s3-deployment/lib/bucket-deployment.ts [hunk 1]
```diff
@@ -235,9 +250,12 @@ export class BucketDeployment extends CoreConstruct {
         SystemMetadata: mapSystemMetadata(props),
         DistributionId: props.distribution?.distributionId,
         DistributionPaths: props.distributionPaths,
+        // Passing through the ARN sequences dependencees on the deployment
+        PassThroughBucketArn: props.destinationBucket.bucketArn,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-s3-deployment/lib/bucket-deployment.ts:254]: Can't we just use the existing `DestinationBucketName` to do this sequencing? Won't it have the same effect, without changing the customer's template?

File: packages/@aws-cdk/aws-s3-deployment/lib/bucket-deployment.ts [hunk 2]
```diff
@@ -235,9 +250,12 @@ export class BucketDeployment extends CoreConstruct {
...
       },
     });
 
+    this.deployedBucket = s3.Bucket.fromBucketArn(this, 'DestinationBucket', Token.asString(deployment.getAtt('DestinationBucketArn')));
```
Comment (skinny85) [inline packages/@aws-cdk/aws-s3-deployment/lib/bucket-deployment.ts:258]: Pretty sure we're missing a unit test for actually using this property in another construct.

**Review by skinny85 — APPROVED** [review]
LGTM. Two minor comments.

File: packages/@aws-cdk/aws-s3-deployment/lib/bucket-deployment.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-s3-deployment/lib/bucket-deployment.ts:254]: I'm not sure I follow. If it doesn't change the template, it won't enforce sequencing, right?

Right now, the dependencies look like this, which lead to a race condition:

```
                                       ┌────────────────────┐
                     { Ref: Bucket }   │                    │
                          ┌────────────│     Deployment     │
                          │            │                    │
┌────────────┐            │            └────────────────────┘
│            │            │                                  
│   Bucket   │◀───────────┤                                  
│            │            │                                  
└────────────┘            │            ┌────────────────────┐
                          │            │                    │
                          └────────────│   BucketConsumer   │
                     { Ref: Bucket }   │                    │
                                       └────────────────────┘
```


We want them to look like this:

```
                                              { GetAtt: [..., 'BucketArn'] }               
┌────────────┐                 ┌────────────────────┐                ┌────────────────────┐
│            │ { Ref: Bucket } │                    │                │                    │
│   Bucket   │◀────────────────│     Deployment     │◀───────────────│   BucketConsumer   │
│            │                 │                    │                │                    │
└────────────┘                 └────────────────────┘                └────────────────────┘
```

Comment (skinny85) [reply packages/@aws-cdk/aws-s3-deployment/lib/bucket-deployment.ts:254]: I meant template changes related to the new `PassThroughBucketArn` property that you added, regardless of whether your customers use the new `deployedBucket ` property or not.

If you used the existing `DestinationBucketName` property for this purpose, I believe existing customers would not have their template affected.

Comment (rix0rrr) [reply packages/@aws-cdk/aws-s3-deployment/lib/bucket-deployment.ts:254]: Oh I see. Is that good enough though? What if the bucket is in another account?

Comment (skinny85) [reply packages/@aws-cdk/aws-s3-deployment/lib/bucket-deployment.ts:254]: If it's in another account, sequencing within the template with `Fn::GetAtt` is irrelevant anyway 🙂.

Comment (rix0rrr) [reply packages/@aws-cdk/aws-s3-deployment/lib/bucket-deployment.ts:254]: That is true, using the bucket still shouldn't fail.

**Pull request merged**